### PR TITLE
added lecture 15

### PR DESCRIPTION
**Pull request by vinaykale64**

### CODE REVIEW

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+  <body>
+    <textarea id="source">
+
+class: center, middle
+
+### W4995 Applied Machine Learning
+
+# Working with Imbalanced Data
+
+03/20/17
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: 02/28/18

**Comment by amueller** [comment]
can you please rename it aml-13?

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 2]
```diff
@@ -0,0 +1,750 @@
...
+
+---
+
+# Changing Thresholds
+
+.smaller[
+```python
+# logistic regression on breast cancer, but change threshold
+data = load_breast_cancer()
+X_train, X_test, y_train, y_test = train_test_split(data.data, data.target, stratify=data.target, random_state=0)
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: needs line break

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+---
+# Mammography Data
+
+.smaller[
+.wide-left-column[
+```python
+import openml
+# mammography dataset https://www.openml.org/d/310
+data = openml.datasets.get_dataset(310)
+X, y = data.get_data(target=data.default_target_attribute)
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: needs line break.

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 4]
```diff
@@ -0,0 +1,750 @@
...
+
+# Mammography Data
+
+```python
+from sklearn.linear_model import LogisticRegression
+scores = cross_val_score(LogisticRegression(),
+                         X_train, y_train, cv=10, scoring='roc_auc')
+print(scores.mean())
+```
+0.919622716696
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: please round to 3 decimals

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 5]
```diff
@@ -0,0 +1,750 @@
...
+```
+0.916512922589
+```python
+undersample_pipe_rf = make_imb_pipeline(RandomUnderSampler(), RandomForestClassifier())
+scores = cross_val_score(undersample_pipe_rf, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.942668403733
+
+- As accurate with fraction of samples!
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: move bullet points to slide notes, add line breaks, round decimals.

**Comment by vinaykale64** [comment]
Sure! Are there any more changes before I push with renaming and changed date ?

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 6]
```diff
@@ -0,0 +1,750 @@
...
+???
+
+---
+
+# Random Oversampling
+
+.smaller[
+```python
+oversample_pipe = make_imb_pipeline(RandomOverSampler(), LogisticRegression())
+scores = cross_val_score(oversample_pipe, X_train, y_train, cv=10, scoring='roc_auc')
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: add line breaks

**Comment by amueller** [comment]
adding review comments right now :)

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 7]
```diff
@@ -0,0 +1,750 @@
...
+]
+
+???
+
+---
+
+# ROC Curves for LogReg
+
+.center[
+![:scale 85%](images/roc_curves_for_logreg.png)
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: can you redo the plot from the notebook and add a title so it's aligned with the plot on the next slide?

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 8]
```diff
@@ -0,0 +1,750 @@
...
+```python
+from sklearn.linear_model import LogisticRegression
+scores = cross_val_score(LogisticRegression(class_weight='balanced'),
+                         X_train, y_train, cv=10, scoring='roc_auc')
+print(scores.mean())
+```
+0.917567920152
+```python
+from sklearn.ensemble import RandomForestClassifier
+scores = cross_val_score(RandomForestClassifier(n_estimators=100, class_weight='balanced'),
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: line break

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 9]
```diff
@@ -0,0 +1,750 @@
...
+![:scale 100%](images/edited_nearest_neighbour_2.png)
+]
+]
+
+???
+
+---
+
+```python
+enn_pipe = make_imb_pipeline(EditedNearestNeighbours(n_neighbors=5),
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: line breaks or smaller font?

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 10]
```diff
@@ -0,0 +1,750 @@
...
+.center[
+![:scale 100%](images/edited_condensed_nn.png)
+]
+
+???
+
+---
+
+```python
+cnn_pipe = make_imb_pipeline(CondensedNearestNeighbour(), LogisticRegression())
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: line breaks or smaller fonts?

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 11]
```diff
@@ -0,0 +1,750 @@
...
+- Leads to very large datasets (oversampling)
+- Can be combined with undersampling strategies
+
+
+???
+
+---
+
+.center[
+![:scale 100%](images/smote.png)
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: can you maybe add "original" and "smote" like in the next slide? Maybe redoing the plots form the notebook might be easiest and labeling there?

File: slides/aml-15-022818-resampling-imbalanced-data/index.html [hunk 12]
```diff
@@ -0,0 +1,750 @@
...
+.right-column[
+smote
+]
+
+???
+
+---
+
+```python
+smote_pipe = make_imb_pipeline(SMOTE(), LogisticRegression())
```
Comment (amueller) [inline slides/aml-15-022818-resampling-imbalanced-data/index.html]: line breaks or font size.

**Review by amueller — COMMENTED** [review]
Looks great. Please make sure that the code is not cut off on any of the slides and also round all numbers to 3 decimals.

**Comment by vinaykale64** [comment]
Added requested changes

File: slides/aml-13-022818-resampling-imbalanced-data/index.html [hunk 13]
```diff
@@ -0,0 +1,774 @@
...
+.smaller[
+```python
+enn_pipe = make_imb_pipeline(EditedNearestNeighbours(n_neighbors=5),
+                             LogisticRegression())
+scores = cross_val_score(enn_pipe, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.920
+```python
+enn_pipe_rf = make_imb_pipeline(EditedNearestNeighbours(n_neighbors
```
Comment (amueller) [inline slides/aml-13-022818-resampling-imbalanced-data/index.html]: this is a horrible line break ;)

File: slides/aml-13-022818-resampling-imbalanced-data/index.html [hunk 14]
```diff
@@ -0,0 +1,774 @@
...
+.smaller[
+```python
+cnn_pipe = make_imb_pipeline(CondensedNearestNeighbour(), LogisticRegression())
+scores = cross_val_score(cnn_pipe, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.919
+```python
+cnn_pipe = make_imb_pipeline(CondensedNearestNeighbour(),
+                          RandomForestClassifier(n_estimators=100))
```
Comment (amueller) [inline slides/aml-13-022818-resampling-imbalanced-data/index.html]: It would be great if you can do pep8 indents (i.e. aligning with the opening parenthesis) for linebreaks if it's reasonable.

File: slides/aml-13-022818-resampling-imbalanced-data/index.html (hunk 13 above)
Comment (vinaykale64) [reply slides/aml-13-022818-resampling-imbalanced-data/index.html]: My bad... will correct it. I was playing with the .smaller[] too, so must have missed it

File: slides/aml-13-022818-resampling-imbalanced-data/index.html (hunk 14 above)
Comment (vinaykale64) [reply slides/aml-13-022818-resampling-imbalanced-data/index.html]: Will do.

File: slides/aml-13-022818-resampling-imbalanced-data/index.html [hunk 15]
```diff
@@ -0,0 +1,774 @@
...
+
+???
+
+
+---
+
+class: center, middle
+
+.center[
+![:scale 100%](images/smote_3.png)
```
Comment (amueller) [inline slides/aml-13-022818-resampling-imbalanced-data/index.html:667]: thanks for recreating the image. We don't really need the one without labels, though (smote_2) and it would be nice to get labels on smote.png as well.

Comment (vinaykale64) [reply slides/aml-13-022818-resampling-imbalanced-data/index.html:667]: Cool. Will do.

**Comment by vinaykale64** [comment]
Added requested changes - 2

**Comment by amueller** [comment]
thanks!
